### PR TITLE
Fix some ARB Extensions core target version & Add Reflections from Gabriel fork.

### LINKED ISSUE #76: Bugs
Hey, first thanks to the creators of this library, I am using the library in my game engine and noticed some bugs:
- When enabling separate shaders option in GLSL 330, with allow extensions 'on', XSC doesn't include 
`#extension GL_ARB_separate_shader_objects : enable` in the generated code.
- Auto binding has many bugs
  - Auto binding takes sampler states into count "when it actually shouldn't" 
         Example HLSL:
         `Texture2D NE_Tex_Diffuse1 : register(t0);`
         `SamplerState NE_Diffuse1_Sampler : register(s0);`
         `Texture2D NE_Tex_Specular1 : register(t1);`
         `SamplerState NE_Specular1_Sampler : register(s1);`
         Generated GLSL
         `layout(binding = 0) uniform sampler2D NE_Tex_Diffuse1;`
         `layout(binding = 2) uniform sampler2D NE_Tex_Specular1; //Should be one`
  - Auto binding doesn't generate binding locations according to DirectX 'register'
         Example HLSL:
         `Texture2D NE_Tex_Diffuse1 : register(t7);`
         `Texture2D NE_Tex_Specular1 : register(t3);`
         Generated GLSL
         `layout(binding = 0) uniform sampler2D NE_Tex_Diffuse1;  //Should be seven`
         `layout(binding = 1) uniform sampler2D NE_Tex_Specular1; //Should be three`

Also a question, Is the library still supported by the creators?
Thanks.

### PR DESCRIPTION
**Pull request by v7medz**
Fixes #76 

Note: Reflections code is taken from https://github.com/Gabriele91/XShaderCompiler/commit/9b0b6feaee47c67288e5ce94beb8128f886814f5

### CODE REVIEW

**Comment by LukasBanana** [comment]
Thanks for you contribution.

I'm ill since this morning and I'm under high time pressure, so it may take one or two weeks until I can review your commits.

**Comment by v7medz** [comment]
No problem, Get well soon!

File: src/Compiler/Xsc.cpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ XSC_EXPORT bool CompileShader(
         reflectionData,
         &timePoints
     );
-
+    #if 0 //no sort
```
Comment (LukasBanana) [inline src/Compiler/Xsc.cpp:37]: Is there a reason this has been disabled? It was originally used for quick command-line feedback of the code reflection, where sorting can be helpful.

File: src/Compiler/ReflectionPrinter.cpp [hunk 2]
```diff
@@ -24,8 +24,18 @@ void ReflectionPrinter::PrintReflection(const Reflection::ReflectionData& reflec
     output_ << R_CodeReflection() << ':' << std::endl;
     indentHandler_.IncIndent();
     {
+        //vector<uniform> -> vector<string>
```
Comment (LukasBanana) [inline src/Compiler/ReflectionPrinter.cpp:27]: There are multiple overloaded versions of the `PrintReflectionObjects` function, which I recommend to overload again instead of creating temporary `vector<string>` containers.

File: src/Compiler/Frontend/GLSL/GLSLExtensions.cpp [hunk 3]
```diff
@@ -31,22 +31,22 @@ const std::map<std::string, int>& GetGLSLExtensionVersionMap()
 
         // ARB
         { E_GL_ARB_arrays_of_arrays,                        430 },
-        { E_GL_ARB_compute_shader,                          110 },
-        { E_GL_ARB_cull_distance,                           110 },
+        { E_GL_ARB_compute_shader,                          430 },
```
Comment (LukasBanana) [inline src/Compiler/Frontend/GLSL/GLSLExtensions.cpp:34]: These versions look good, but I think it would have been better if they are in a separate pull request.

File: src/Compiler/Backend/GLSL/GLSLExtensionAgent.cpp [hunk 4]
```diff
@@ -87,7 +87,7 @@ std::set<std::string> GLSLExtensionAgent::DetermineRequiredExtensions(
 
     if (separateShaders)
     {
-        if (shaderTarget != ShaderTarget::FragmentShader && shaderTarget != ShaderTarget::ComputeShader)
+        if (shaderTarget != ShaderTarget::ComputeShader)
```
Comment (LukasBanana) [inline src/Compiler/Backend/GLSL/GLSLExtensionAgent.cpp:90]: What's the reason for this change? As far as I remember, the original code for *separate shader objects* was introduced by @BearishSun so I'm not sure if this condition is sensible.

File: src/Compiler/AST/Visitor/ReflectionAnalyzer.cpp [hunk 5]
```diff
@@ -148,21 +454,40 @@ IMPLEMENT_VISIT_PROC(BufferDeclStmnt)
...
             {
+                DataType baseType = DataType::Undefined;
                 /* Add variable as uniform */
-                data_->uniforms.push_back(ast->ident);
+                Reflection::Uniform uniform;
+                uniform.ident = ast->ident;
+                uniform.type = Reflection::UniformType::Variable;
+                //find type
+                if (auto basetype_denoter = typeSpecifier->typeDenoter->As<BaseTypeDenoter>())
+                    uniform.baseType = (int)DataTypeToReflType(basetype_denoter->dataType);
```
Comment (LukasBanana) [inline src/Compiler/AST/Visitor/ReflectionAnalyzer.cpp:486]: Please be so good and always use the appropriate C++ type cast. I never use C casts in my C++ projects for several reasons:
1. They can silently cast constness away! This is really evil and I have only exactly three locations where I use `const_cast` (see `Expr::Find` and co) which is documented why it is ok at that point.
2. You cannot always know exactly which kind of cast the compiler will use, since C-style casts are a combination of multiple casts but only one is choosen by the compiler.
3. Searching for `(int)` may also find function declarations with only a single anonymous parameter.

File: inc/Xsc/Reflection.h [hunk 6]
```diff
@@ -124,6 +124,279 @@ struct BindingSlot
...
+    Texture2DArray,
+    Texture3D,
+    TextureCube,
+    TextureCubeArray,
+    Texture2DMS,
+    Texture2DMSArray,
+};
+
+//! Enum of types of a parameter of a function
+enum class VarType
```
Comment (LukasBanana) [inline inc/Xsc/Reflection.h:283]: This enum looks very redundant, since the elements are almost identical to the one in `DataType`. Since I don't really like the idea to have two enums named `DataType` in the same project (one in the private code, and one in the public header files) I think `VarType` should be the preferred name for the code reflection.
It could be used as a common data type enumeration for the code reflection, i.e. it should integrate `DataType` into it.

**Review by LukasBanana — COMMENTED** [review]
It would be great if you could separate your pull request into the one with the new extension versions and the code reflection part.
The code reflection part needs some refinements, and maybe the type system for reflection output needs to be further discussed.

File: src/Compiler/Backend/GLSL/GLSLExtensionAgent.cpp (hunk 4 above)
Comment (v7medz) [reply src/Compiler/Backend/GLSL/GLSLExtensionAgent.cpp:90]: You need to enable separate shaders extension in all shaders except compute shaders in order to work.

Also if you don't opengl will spam errors, and I tested it on my end.

File: src/Compiler/Frontend/GLSL/GLSLExtensions.cpp (hunk 3 above)
Comment (v7medz) [reply src/Compiler/Frontend/GLSL/GLSLExtensions.cpp:34]: IDK how, I am pretty noob in all those git stuff... :)

File: src/Compiler/Xsc.cpp (hunk 1 above)
Comment (v7medz) [reply src/Compiler/Xsc.cpp:37]: As I said all reflection code is copied from @Gabriele91, I know that they need a bit of refinement, I am currently working on normal mapping in my engine when I finish it, I will modify the reflection code and fix the texture binding slot.
